### DISCUSSION TITLE
Support for an Auto-Expire Scaler

### DISCUSSION THREAD

**Discussion by droessmj** [Azure/bellhop#55] [Ideas]
Bellhop seems to be intended for scaling resources up/down across various service tiers. This idea is less of a scaling item, but does tie directly to the intent of cost savings. 

I'd propose the tags `autoExpire-Enabled = true` and `autoExpire-Date = <DateTime UTC>` which could be applied to any resource and which Bellhop would process as deletes once the autoExpire-Date was reached. 

Possible edge-case considerations: 
  - Delete is attempted but fails due to dependent resources not being deleted (e.g., Public IPs and Application Gateways)
  - Linked/associated resources are effectively orphaned during deletion (VM & disk, NSG)

**Comment by CloudViking** [comment]
I like this idea of an `autoExpire` tag. To keep with the project theme, this would be a situation of "set a time and Bellhop will automatically check you out of the hotel...".

What would you think about combining the above tags into a single key-value pair?  Something along the lines of: 
- `autoExpire-Enable = <DateTime UTC>`

My thought being the existence of the `autoExpire-Enable` tag on the resource would enable the action, and the tag value would be the UTC DateTime to remove the resource. Keeping the number of tags Bellhop uses to a minimum could become something to consider moving forward. 

Good call on the potential edge cases around delete failures and dependent resources. Off the top of my head:
- MVP might be good logging and alerting on failed deletes
    - KQL Query/Sample Workbook/Built-in Notifications
    - User is notified
    - User manually deletes and cleans up resources

Definitely open to ideas!

**Reply by DCMattyG** [reply]
Hey @CloudViking, I **really** like the idea of combining the tags! I would be _implied_ that auto-expire was enabled if we just had a single tag which either exists, or does not. Let us know your feedback on this @droessmj and we can get a feature request spun up for this item.

**Reply by DCMattyG** [reply]
Additional thoughts on this....

It would seem to me that the responsibility here would fall to the Scaler, not the Engine. That way, per @droessmj's note above, the edge cases could be accounted for in that:

1) A scaler would need to exist for the resource in order for it to be acted upon in Bellhop.
2) The logic for the linked/associated items could be handled in said scaler.

That's my $0.02 at the moment.
